src/persist-client/src/impl/encoding.rs

```diff
@@ -98,14 +99,46 @@ where
     T: Timestamp + Lattice + Codec64,
     D: Codec64,
 {
-    pub fn decode(buf: &[u8]) -> Result<Self, CodecMismatch> {
+    pub fn decode(build_version: &Version, buf: &[u8]) -> Result<Self, CodecMismatch> {
         let proto = ProtoStateRollup::decode(buf)
             // We received a State that we couldn't decode. This could happen if
             // persist messes up backward/forward compatibility, if the durable
             // data was corrupted, or if operations messes up deployment. In any
             // case, fail loudly.
             .expect("internal error: invalid encoded state");
-        Self::try_from(proto).expect("internal error: invalid encoded state")
+        let state = Self::try_from(proto).expect("internal error: invalid encoded state")?;
+
+        // If persist gets some encoded ProtoState from the future (e.g. two
+        // versions of code are running simultaneously against the same shard),
+        // it might have a field that the current code doesn't know about. This
+        // would be silently discarded at proto decode time. Unknown Fields [1]
+        // are a tool we can use in the future to help deal with this, but in
+        // the short-term, it's best to keep the persist read-modify-CaS loop
+        // simple for as long as we can get away with it (i.e. until we have to
+        // offer the ability to do rollbacks).
+        //
+        // [1]: https://developers.google.com/protocol-buffers/docs/proto3#unknowns
+        //
+        // To detect the bad situation and disallow it, we tag every version of
+        // state written to consensus with the version of code used to encode
+        // it. Then at decode time, we're able to compare the current version
+        // against any we receive and assert as necessary. Initially we reject
+        // any version from the future (most conservative but easiest to reason
+        // about). If/when we support deploy rollbacks and rolling upgrades, we
+        // can adjust this assert as necessary to reflect the policy.
```

Comment:
wouldn't mind highlighting the current policy + how it'll need to change in the future in a separate paragraph

Reply:
done!